src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenSpanBasedStringConcatTests.cs

```diff
@@ -0,0 +1,4122 @@
+            //     static string M2(string s, char c) => c + s;
+            Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "c").WithArguments("System.Object", "ToString").WithLocation(14, 43));
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/66827")]
+    public void ConcatTwo_CharDoesntOverrideObjectToString()
+    {
+        var corlib_cs = """
+            namespace System
+            {
+                public class Object
+                {
+                    public virtual string ToString() => null;
+                }
+                public class String
+                {
+                    public static string Concat(string str0, string str1) => null;
+                    public static string Concat(ReadOnlySpan<char> str0, ReadOnlySpan<char> str1) => null;
+                    public static implicit operator ReadOnlySpan<char>(string value) => default;
+                }
+                public class ValueType { }
+                public struct Char { }
+                public struct Void { }
+                public struct Int32 { }
+                public struct Byte { }
+                public struct Boolean { }
+                public struct ReadOnlySpan<T>
+                {
+                    public ReadOnlySpan(ref readonly T reference) { }
+                }
+                public class Enum : ValueType { }
+                public class Attribute { }
+                public enum AttributeTargets { }
+                public class AttributeUsageAttribute : Attribute
+                {
+                    public AttributeUsageAttribute(AttributeTargets validOn) { }
+
+                    public bool AllowMultiple { get { return default; } set { } }
+                    public bool Inherited { get { return default; } set { } }
+                }
+            }
+            """;
+
+        var corlib = CreateEmptyCompilation(corlib_cs).EmitToImageReference();
+
+        var source = """
+            public class Test
+            {
+                static string M(string s, char c) => s + c;
+            }
+            """;
+
+        var comp = CreateEmptyCompilation(source, [corlib]);
+        comp.VerifyDiagnostics();
+
+        var verifier = CompileAndVerify(compilation: comp, verify: RuntimeUtilities.IsCoreClr8OrHigherRuntime ? default : Verification.Skipped);
+        verifier.VerifyDiagnostics();
+
+        // No matter whether `char` directly overrides `ToString` or not, we still produce span-based concat if we can
+        verifier.VerifyIL("Test.M", """
+            {
+              // Code size       21 (0x15)
+              .maxstack  2
+              .locals init (char V_0)
+              IL_0000:  ldarg.0
+              IL_0001:  call       "System.ReadOnlySpan<char> string.op_Implicit(string)"
+              IL_0006:  ldarg.1
+              IL_0007:  stloc.0
+              IL_0008:  ldloca.s   V_0
+              IL_000a:  newobj     "System.ReadOnlySpan<char>..ctor(ref readonly char)"
```

Comment:
For whatever reason with my custom-made corlib in IL baseline constructor argument is displayed as `ref readonly char` while with real corlib it is `in char`. @AlekseyTs Is it ok?

Reply:
> For whatever reason with my custom-made corlib in IL baseline constructor argument is displayed as `ref readonly char` while with real corlib it is `in char`. Is it ok?

It looks like it matches the way you declared the constructor. Our ref assemblies are probably declaring the parameter as `in`. It is quite possible it will be changed to `ref readonly` in new version of BCL. It is up to you if you want to align with 'in' or not.

Reply:
Ok, gonna leave it as is then